app/scripts/lib/snap-keyring/metrics.ts

```diff
@@ -0,0 +1,51 @@
+import MetamaskController from '../../metamask-controller';
+
+export default async function getSnapAndHardwareInfoForMetrics(
+  metamaskController: MetamaskController,
+) {
+  if (!metamaskController?.preferencesController) {
+    return {}; // if it's coming from a unit test, there's no metamaskController or preferencesController
+  }
```

Comment:
Why not only pass the necessary methods from instead of the entire `metamaskController` object?

Reply:
Well it does call five different methods from `metamaskController` -- `getSelectedAddress` `getTruncated` `getKeyringForAccount` `getAccountType` and `getDeviceModel`.  It seemed cleaner to get them all from this one file, rather than access and pass them all externally from the Transaction route and the Signature route.

Reply:
I would prefer to not pass the entire `metamaskController` and pass only what is needed.

Reply:
i think we can pass in a controller messenger instead of passing the entire `metamaskController`

we can construct a messenger like this and pass it into `createRPCMethodTrackingMiddleware`. With this we can get the snaps keyring, get the keyring for the account the snaps data. 
```javascript
    const rpcControllerMessenger = this.controllerMessenger.getRestricted({
      name: 'RPCMethodMessenger',
      allowedActions: [
        'KeyringController:getKeyringForAccount',
        'KeyringController:getKeyringByType',
        'SnapController:getSubjectMetadata',
      ],
    });
```

to call the action, it would be something like this: 
```javascript
const [snapKeyring] = await this.controllerMessenger.call(
      'KeyringController:getKeyringsByType',
      'Snap Keyring',
    );
```

For the selectedAddress, we can just pass a call back to that method instead of passing the entire preferencesController.

```javascript
engine.push(
      createRPCMethodTrackingMiddleware({
        trackEvent: this.metaMetricsController.trackEvent.bind(
          this.metaMetricsController,
        ),
        getMetricsState: this.metaMetricsController.store.getState.bind(
          this.metaMetricsController.store,
        ),
        securityProviderRequest: this.securityProviderRequest.bind(this),
        controllerMessenger: rpcControllerMessenger,
        getSelectedAddress: this.preferencesController.getSelectedAddress.bind(
          this.preferencesController,
        ),
      }),
    );
```